src/execution/operator/aggregate/physical_streaming_window.cpp

```diff
@@ -420,15 +478,70 @@ OperatorResultType PhysicalStreamingWindow::Execute(ExecutionContext &context, D
 	for (idx_t col_idx = 0; col_idx < input.data.size(); col_idx++) {
 		chunk.data[col_idx].Reference(input.data[col_idx]);
 	}
-	// Compute window function
-	const idx_t count = input.size();
+	idx_t count = input.size();
+
+	//	Handle LEAD
+	if (state.lead_count > 0) {
+		//	Break the combined delayed + input into chunk + delayed
+		auto &delayed = state.delayed;
+		const idx_t available = delayed.size() + count;
+		//	If we don't have enough to produce a single row,
+		//	then just delay more rows and return nothing
+		if (available <= state.lead_count) {
+			delayed.Append(input, true);
+			chunk.SetCardinality(0);
+			return OperatorResultType::NEED_MORE_INPUT;
+		}
+
+		const idx_t prefixed = MinValue<idx_t>(delayed.size(), count);
+		if (prefixed) {
+			//	Copy delayed[:prefixed] => chunk[:prefixed]
+			chunk.Reset();
+			for (idx_t col_idx = 0; col_idx < delayed.data.size(); col_idx++) {
+				auto &src = delayed.data[col_idx];
+				auto &dst = chunk.data[col_idx];
+				VectorOperations::Copy(src, dst, prefixed, 0, 0);
+			}
+			const idx_t ragged = delayed.size() - prefixed;
+			if (ragged) {
+				// Corner case: didn't consume all the delayed values
+				// Example: input has 1 row but delay is 2 rows.
+				// Copy delayed[ragged:] => delayed[0:]
+				auto &shifted = state.shifted;
+				shifted.Reset();
+				delayed.Copy(shifted, ragged);
+				delayed.Reset();
+				shifted.Copy(delayed, 0);
+				// Copy input[:] => delayed[ragged:]
+				delayed.Append(input, true);
+			} else {
+				// Copy input[:count - prefixed] => chunk[prefixed:count]
+				const idx_t src_count = count - prefixed;
+				for (idx_t col_idx = 0; col_idx < input.data.size(); col_idx++) {
+					auto &src = input.data[col_idx];
+					auto &dst = chunk.data[col_idx];
+					VectorOperations::Copy(src, dst, src_count, 0, prefixed);
+				}
+				// Copy input[count - prefixed:] => delayed
+				delayed.Reset();
+				input.Copy(delayed, src_count);
```

Comment:
In this implementation we are copying almost all of the rows, as we are essentially doing:


* rows `[0..2048)` -> emit `[0..2047)`, copy `[2047..2048)` into delayed
* rows `[2048..4096)`-> copy delayed into chunk, copy `[2048..4095)` into chunk, copy `[4095..4096)` into delayed
* etc

An alternative implementation would be to make use of the `OperatorResultType::HAVE_MORE_OUTPUT` to emit partial chunks. When `OperatorResultType::HAVE_MORE_OUTPUT` is returned, we are called again with the same input chunk. For example we could do something like this:

* rows `[0..2048)` -> emit `[0..2047)`, copy `[2047..2048)` into delayed
* rows `[2048..4096)` -> emit delayed, return `OperatorResultType::HAVE_MORE_OUTPUT`
* rows `[2048..4096)` -> emit `[2048..4095)`, copy `[4095..4096)` into delayed
* etc

This way we only copy `lead_count` rows. The trade-off is that we emit more (and smaller) chunks. Whether or not this is better - I don't know :) I suspect it also depends on the size of the payload and the data types, e.g. if we have strings or more complex types like lists copying is more expensive versus only copying integers as is currently done in the benchmark. Might be worth investigating.

Reply:
This was bugging me when I wrote it (and I think the main reason it took me so long was I was trying desperately to avoid copying!)

I think your instinct about common use cases is correct, so we would be emitting one block rows all the time, which would make us into a row store!  Our threshold in `CachingOperatorState` is 64 so we already want larger blocks sometimes, but by not compacting ourselves, we would end up alternating and COS would not be able to do anything (unless it did precisely what we were trying to avoid...)

Reply:
That's true - but we would be emitting an alternating `1  row-> 2047 rows -> 1 row -> 2047 rows`, which from a performance perspective should be equivalent to emitting ~1024 rows per vector. That's a downgrade but not terrible. The caching operator is mainly to avoid worst case scenarios where we emit 1 row at a time.

That said it's definitely a drawback - I'm not sure which would be better. I suspect it's payload dependent - the current approach might win in case of a small payload, while avoiding copying would win in case of a larger one.

Reply:
I can take a look, but do we want to treat this as further work?